pkg/cri/server/helpers_linux.go

```diff
@@ -161,7 +162,7 @@ func openLogFile(path string) (*os.File, error) {
 // unmountRecursive unmounts the target and all mounts underneath, starting with
 // the deepest mount first.
 func unmountRecursive(ctx context.Context, target string) error {
-	mounts, err := mount.Self()
+	mounts, err := mountinfo.GetMounts(nil)
```

Comment:
Two things.

1. This might benefit from using PrefixFilter (and when the logic will be simplified -- there's no need to do filepath.Rel and strings.HasPrefix below).

2. Not really related to this, but there's a little known fact that umount(MNT_DETACH) is actually recursive in Linux, IOW this function can be replaced with `unix.Umount(target, unix.MNT_DETACH)` (or `mount.UnmountAll(target, unix.MNT_DETACH)`) -- provided that `target` itself is a mount point. Related: https://github.com/moby/sys/pull/26

3. Definitely not related to this, but I took a look at `pkg/mount.UnmountAll` and :open_mouth: it retries umount without a delay on any error rather than EINVAL forever. This means, for example, if someone denies umount syscall using seccomp, we'll end up calling umount in a busy loop indefinitely.

Reply:
@kolyshkin would that mean we can replace this whole function with `moby/mount.RecursiveUnmount()`, or would we still need some of the logic below?

Reply:
> it retries umount without a delay on any error rather than EINVAL

I see it returns on errors (only returns `nil` if it was an `EINVAL`), or is that not the part you were looking at? https://github.com/containerd/containerd/blob/45d8a7e4a2d1e69dad4e1d879a7a7210a6d7ec71/mount/mount_linux.go#L179-L182

Reply:
> I see it returns on errors (only returns nil if it was an EINVAL), or is that not the part you were looking at?

You're right, I misread the code in there :) so item 3 above can be crossed out.

Re. item 2, I finally understood the logic of containerd's `UnmountAll()` (it works around overmounts, i.e. when the same mountpoint being mounted to multiple times) and I think we should do something like this in moby/sys/mount, too. In the meantime, let's leave this as is, IOW ACK.